### DISCUSSION TITLE
Dont see full crate path of nvim_cmp suggestions

### DISCUSSION THREAD

**Discussion by poctek** [mrcjkb/rustaceanvim#414] [Q&A]
Hi everyone! I've recently switched from rust-tools and having some problems with configuration. The main is that for some reason it seems that rust analyzer doesn't provide full create path of a suggestion anymore (like `use some_crate::U256`) which leads to a problem when I have multiple structs with the name from different crates in suggestions like on the screenshot attached.

<img width="445" alt="Screenshot 2024-05-31 at 00 40 08" src="https://github.com/mrcjkb/rustaceanvim/assets/9880482/b69dd251-395a-4674-94a1-5a41f3633950">

Earlier it used to look like this:

<img width="445" alt="Screenshot 2024-05-31 at 00 40 08 2" src="https://github.com/mrcjkb/rustaceanvim/assets/9880482/8206a021-d361-4951-8f2b-6d6ae28b9374">

I've already remove the configuration for the plugin at all, so right now all I have related to rust in my init.lua are

```
Plug 'mrcjkb/rustaceanvim'
Plug 'rust-lang/rust.vim'
```

Maybe anyone here can give me a direction on where to dig or how to fix this issue?

**Comment by mrcjkb** [comment]
Hey :wave: 

This plugin doesn't do anything related to auto completion besides register the client capabilities of plugins it detects are installed:

https://github.com/mrcjkb/rustaceanvim/blob/2fa45427c01ded4d3ecca72e357f8a60fd8e46d4/lua/rustaceanvim/config/server.lua#L150-L155

Your issue is most likely one of the following:

- You're using a different version of rust-analyzer (For example, this plugin doesn't pick up mason.nvim installs (see `:h rustaceanvim.mason`).
- Or you're missing a rust-analyzer server configuration.

**Reply by poctek** [reply]
Yep, I understand that something is wrong with my rust-analyzer config, maybe someone here knew what's missing in my config and could provide the settings I need. As far I as see, there're no conflicts between rust analyzer and the plugin

```
==============================================================================
rustaceanvim: require("rustaceanvim.health").check()

Checking for Lua dependencies ~
- WARNING dap not installed. Needed for debugging features [mfussenegger/nvim-dap](https://github.com/mfussenegger/nvim-dap)

Checking external dependencies ~
- OK rust-analyzer: found rust-analyzer 0.0.0 (71a816a90 2024-05-26)
- OK Cargo: found cargo 1.78.0 (54d8815d0 2024-03-26)
- OK rustc: found rustc 1.78.0 (9b00956e5 2024-04-29)

Checking config ~
- OK No errors found in config.

Checking for conflicting plugins ~
- OK No conflicting plugins detected.

Checking for tree-sitter parser ~
- WARNING No tree-sitter parser for Rust detected. Required by 'Rustc unpretty' command.

```

**Reply by jaboatman** [reply]
I also ran into this. Not sure where the issue is. I switched from rust-tools to rustaceanvim and the label details are missing from completion. I have tried multiple versions of rust-analyzer to no avail.

I went through all of the source code of rust-tools and this project to compare LSP configuration and I could not find any reason why this issue would start happening.

**Reply by jaboatman** [reply]
I went back to my rust-tools setup and was able to reproduce it there by attaching default lsp_properties from `cmp_nvim_lsp`. Suggesting that if we could have a way for rustaceanvim to _not_ include those defaults it might work. Short of removing the plugin though, I don't see a way to force it not to include it. Really, it should be opt-in in my opinion.

**Reply by jaboatman** [reply]
Turns out it still isn't that simple. I replaced the capabilities with those set by `require('rust-tools').setup {}` and it still doesn't work. Same version of RA. Maybe somewhere some RA config is being set that I'm not finding.

**Reply by mrcjkb** [reply]
The best thing to do in those cases is to bisect your plugins/config.

**Reply by jaboatman** [reply]
I commented out the following line:
```lua
  return vim.tbl_deep_extend(
    'force',
    rs_capabilities,
    -- cmp_capabilities, -- here
    selection_range_capabilities,
    folding_range_capabilities
  )
```

And that fixed the issue. So having cmp causes the issue. I'm not sure what specifically in there causes it.

**Comment by jaboatman** [comment]
OK, I finally tracked this down.

* If you are using `nvim-cmp`
* and if you have overwritten its format function to replace the `item.menu` field

then you will see the above issue. This is because it sets the field `capabilities.textDocument.completion.completionItem.labelDetailsSupport = true`, which `nvim-cmp` will then put the label details (which include the import path) into the `item.menu` field on completion. If this configuration value is not provided, then rust-analyzer will format the completion results to include the import path in the result string instead.

However, even if you override the configuration to `false`, it still doesn't work, due to a bug in rust-analyzer: https://github.com/rust-lang/rust-analyzer/pull/18388

So to fix it, do one of the following:
 
* manually remove using `cmp_capabilities` as in the above (not recommended)
* don't override `cmp` `format` `item.menu` field (or reformat it to explicitly use the label details)
* use my fork of rust-analyzer and override `capabilities.textDocument.completion.completionItem.labelDetailsSupport = false` (not recommended)
